### PR TITLE
Hexagon and truncated hexagon

### PR DESCRIPTION
**Pull request by ChasingNeutrons**
I've developed an axis-parallel hexagon. It is regular in all directions, i.e., has the same width. Its main limitation is that it can't handle reflective boundaries. This is due to the difficulty of handling reflective boundaries by transform. However, I imagine for most cases of interest where this is used as a boundary that periodic boundaries will be sufficient. I have added unit tests and modified the documentation. I have also added a truncated hexagon, which does allow any boundary conditions along the perpendicular direction.

### CODE REVIEW

File: docs/Input Manual.rst [hunk 1]
```diff
@@ -782,6 +782,30 @@ Example: ::
...
 
+
+* hexagon: hexagon oriented parallel to a given axis. The orientation is chosen by specifying 
+  ``xHexagon``, ``yHexagon`` or ``zHexagon``. The hexagon can be either pointy-topped or flat-topped.
+  This orientation in the plane is determined by specifying either ``orientation 1;`` or ``orientation 2;``.
+  Choosing ``1`` is pointy: the hexagon points towards the second of the two axis in the plane (yZ, xZ, or
+  xY for the three types, respectively) while the first axis is perpendicular to the flat surfaces. Likewise,
+  choosing ``2`` is flat-topped. Can have periodic and vacuum boundaries, but not reflective.
+
+  - origin: (x y z) position of the centre of the hexagon. [cm]
```
Comment (valeriaRaffuzzi) [inline docs/Input Manual.rst]: Add: 'the entry in the axis along the cylinder direction is ignored'.

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+  end function myType
+
+  !!
+  !! Initialise hexagon from a dictionary
+  !!
+  !! See surface_inter for more details
+  !!
+  !! Errors:
+  !!  fatalError for -ve id or meaningful halfspace.
+  !!  fatalError for unrecognised oritnetation
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: typo ``orientation``

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+
+  end function myType
+
+  !!
+  !! Initialise hexagon from a dictionary
+  !!
+  !! See surface_inter for more details
+  !!
+  !! Errors:
+  !!  fatalError for -ve id or meaningful halfspace.
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: What does ``meaningful halfspace`` mean?

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+    class(dictionary), intent(in)            :: dict
+    integer(shortInt)                        :: id, N, i, orient
+    real(defReal), dimension(:), allocatable :: temp
+    character(nameLen)                       :: type
+    real(defReal), dimension(6)              :: angles
+    character(100), parameter :: Here = 'init (hexagon_class.f90)'
+
+    ! Load id
+    call dict % get(id,'id')
+    if (id <= 0) call fatalError(Here, 'ID must be <= 0. Is: '//numToChar(id))
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: Type ``ID must be >0``

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+
+  !!
+  !! Returns TRUE if particle is going into +ve halfspace
+  !!
+  !! See surface_inter for details
+  !!
+  !! Works by:
+  !!  1) Determine a plane in which direction is the closest
+  !!  2) Use normal for this plane and project distance
+  !!  3) Determinie halfspace based on sign of the projection
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: Typo: ``determine``

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+  pure function normal(self, r, u) result(n)
+    class(hexagon), intent(in)              :: self
+    real(defReal), dimension(3), intent(in) :: r
+    real(defReal), dimension(3), intent(in) :: u
+    real(defReal), dimension(3)             :: n
+    real(defReal), dimension(2)             :: rl, e, nTrial, nBest
+    real(defReal)                           :: d, dist
+    integer(shortInt)                       :: i
+    
+    ! Get position in the plane & direction
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: Just position, remove ``& direction``

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+  !!
+  !! See surface_inter for details
+  !!
+  subroutine setBC(self, BC)
+    class(hexagon), intent(inout)               :: self
+    integer(shortInt), dimension(:), intent(in) :: BC
+    character(100),parameter                    :: Here = 'setBC (hexagon_class.f90)'
+
+    if(size(BC) /= 6) call fatalError(Here, 'Wrong size of BC string. Must be 6')
+    if(all(BC /= BC(1))) call fatalError(Here, 'All BCs must be identical')
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: This is confusing because what does VACUUM BCs mean for an infinite cylinder? 
Maybe just check the +/- values of self % plane.

File: Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90 [hunk 8]
```diff
@@ -0,0 +1,640 @@
...
+  subroutine explicitBC(self, r, u)
+    class(hexagon), intent(in)                 :: self
+    real(defReal), dimension(3), intent(inout) :: r
+    real(defReal), dimension(3), intent(inout) :: u
+    integer(shortInt)                          :: i, iPlane
+    real(defReal)                              :: dist, proj, maxProj
+    real(defReal), dimension(2)                :: rl, e, v1, v2, n, nBest
+    character(100), parameter :: Here = 'explicitBC (hexagon_class.f90)'
+
+    ! Get position in the plane & direction
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/hexagon_class.f90]: remove ``& direction``

File: Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90 [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+  !!
+  !! Origin entry in axis along the cylinder direction is ignored.
+  !! Halfwidth is half the flat-to-flat distance.
+  !! Halfheight is half the top-to-bottom distance.
+  !!
+  !! Boundary Conditions:
+  !!   Only a single BC value within the hexagon plane.
+  !!   Any BC can be used along the axial direction.
+  !!
+  !!   For now, each radial face must have the same BC. Supports either periodic or vacuum BCs.
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90]: Redundant comments, maybe a leftover?
Only a single.....
For now, each radial face....

File: Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90 [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+  !!   origin      -> position of the middle of the hexagon
+  !!   halfwidth   -> Halfwidth of the hexagon from one flat to the opposite. Same in all directions
+  !!   halfheight  -> Halfheight of the hexagon from the top to the bottom.
+  !!   BC          -> Boundary conditions flag for all faces in the plane and on the top and bottom.
+  !!   plane       -> Indices of the axis, which are the plane of the cylinder
+  !!                  (e.g. for xTruncHexagon, y-z)
+  !!   axis        -> Index of the axis of the plane
+  !!   flatAxis    -> Index into plane, deciding which plane axis has the flat of the hexagon
+  !!   pointAxis   -> Index into plane, deciding which plane axis has the point of the hexagon
+  !!   verts       -> Vertices of the hexagon
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90]: vertices on the 2D plane*

File: Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90 [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+  end function myType
+
+  !!
+  !! Initialise truncated hexagon from a dictionary
+  !!
+  !! See surface_inter for more details
+  !!
+  !! Errors:
+  !!  fatalError for -ve id or meaningful halfspace.
+  !!  fatalError for unrecognised oritnetation
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90]: As above:
meaningful halfspace?
orientation*

File: Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90 [hunk 12]
```diff
@@ -0,0 +1,791 @@
...
+
+  !!
+  !! Returns TRUE if particle is going into +ve halfspace
+  !!
+  !! See surface_inter for details
+  !!
+  !! Works by:
+  !!  1) Determine a plane in which direction is the closest
+  !!  2) Use normal for this plane and project distance
+  !!  3) Determinie halfspace based on sign of the projection
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90]: Determine

File: Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90 [hunk 13]
```diff
@@ -0,0 +1,791 @@
...
+  subroutine explicitBC(self, r, u)
+    class(truncHexagon), intent(in)            :: self
+    real(defReal), dimension(3), intent(inout) :: r
+    real(defReal), dimension(3), intent(inout) :: u
+    integer(shortInt)                          :: i, iPlane, bc
+    real(defReal)                              :: dist, proj, maxProj, rz
+    real(defReal), dimension(2)                :: rl, e, v1, v2, n, nBest
+    character(100), parameter :: Here = 'explicitBC (truncHexagon_class.f90)'
+
+    ! Get position in the plane & direction
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/CompositeSurfaces/truncHexagon_class.f90]: remove ``& direction``

File: Geometry/Surfaces/Tests/hexagon_test.f90 [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+  function newTestCase(param) result(tst)
+    type(hexParam), intent(in)  :: param
+    type(test_hexagon)          :: tst
+    type(dictionary)            :: dict
+    character(nameLen)          :: type
+    real(defReal), dimension(3) :: origin
+    real(defReal)               :: hw
+    integer(shortInt)           :: orient
+
+    ! Select type of squareCylinder and axis
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/Tests/hexagon_test.f90]: comment: type of hexagon

File: Geometry/Surfaces/Tests/hexagon_test.f90 [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+    call this % surf % transformBC(r, u)
+    @assertEqual(r_ref, r, TOL)
+    @assertEqual(u_ref, u, TOL)
+
+    ! Reflective
+    !BC = REFLECTIVE_BC
+    !call this % surf % setBC(BC)
+
+    ! Shift left by one lattice starting just off centre
+    !r([ax, p1, p2]) = [ZERO, -5.5_defReal, 1.5_defReal]
```
Comment (valeriaRaffuzzi) [inline Geometry/Surfaces/Tests/hexagon_test.f90:353]: Are these commented lines leftover from squareCylinder, or is the test actually correct and just waiting for the implementation to be uncommented?

**Review by valeriaRaffuzzi — APPROVED** [review]
All looks good! I am approving it despite some comments because they are all related to the documentation and comments in the code, rather than changes or suggestions to the code itself. So feel free to merge directly after fixing the comments :) 

Thanks for the implementation!

File: Geometry/Surfaces/Tests/hexagon_test.f90 (hunk 15 above)
Comment (ChasingNeutrons) [reply Geometry/Surfaces/Tests/hexagon_test.f90:353]: They are hopefully ready to go when the transformBC is implemented. I've put a TODO over them.
